### ISSUE TITLE
Add ability to control feature altitude via a style spec property

### ISSUE THREAD

**Issue by cigone-openindoor** [maplibre/maplibre-gl-js#644]
Clone this issue: https://github.com/mapbox/mapbox-gl-js/issues/3993

The main idea is to provide the ability to have symbol (marker, images, text) at a given elevation.

It would be convenient to mix with extrusion capabilities, like for 3D buildings.

**Comment by yangjuncode** [comment]
this is the  painpoint when using  extrusion

**Comment by leehoss** [comment]
I second this feature ⭐

**Comment by dechristopher** [comment]
Something like this would be useful for visualizing 3D paths as well. For example lines between two points in 3D space after #165 is implemented, or an implementation that allows for 3D measurement like Google Earth Pro does. There are many use cases like this and I think it would be a welcome addition from both GIS pros and users alike.

**Comment by spencerthayer** [comment]
Has there been any progress getting markers, points, or any object to render with a z-offset or altitude?

**Comment by samuel-git** [comment]
I would suggest using a third element in position arrays of the source, as defined in GeoJSON spec.

> An OPTIONAL third position element SHALL be the height in meters above the WGS 84 reference ellipsoid.
https://datatracker.ietf.org/doc/html/draft-butler-geojson#section-3

This could be a starting point for 3D LineStrings, Points and maybe Polygons as well. For LineStrings, the third position element would make rich 3D Lines possible with more than two points (e.g. aircraft flight paths) opposed to fixed-height extrusion in style spec.

@wipfli maybe you can point me to the right parts of the code and I can give this a try. But I am a beginner with WebGL.

**Comment by HarelM** [comment]
I would look at the line_program related code.

**Comment by wipfli** [comment]
Oh this feature would be so cool. @samuel-git if you implement something like this I will immediately test it with my hot-air balloon GPS tracks...

**Comment by cns-solutions-admin** [comment]
We have to display aircraft tracks and on a 3d map having them on the ground and vanishing behind mountains is not nice.

I definitely vote for using the optional 3rd geojson coordinate item, as it would also allow 3D-lines (e.g. the planned flight route or the actual flight path of aircrafts.

**Comment by leehoss** [comment]
Been waiting for this feature to help with indoor multilevel routing …. This would be a game changer , please let’s get momentum on this feature :)

**Comment by NFmax** [comment]
It would be awesome to have the addressed feature implemented! However, it might be a feasible alternative for now to use [custom style layers](https://maplibre.org/maplibre-gl-js-docs/example/custom-style-layer/) for some of the use cases named above. 

Here is a simple example of how that could look like: [https://github.com/weglide/maplibre-3d-linestring-demo](https://github.com/weglide/maplibre-3d-linestring-demo)
<img width="1599" alt="image" src="https://raw.githubusercontent.com/weglide/maplibre-3d-linestring-demo/main/screenshot.png">

I am a WebGL beginner, and I assume the performance of this code is nowhere near its optimum, but it proves the possibility and might help some of you, too. Of course, feedback is welcome if someone has ideas for improvements :)

**Comment by HarelM** [comment]
@NFmax can you post your demo using github pages? I'm curious as to how this would look/interact...

**Comment by NFmax** [comment]
It should be possible to run this locally without any effort. Just download the .html file and open it with a browser ;)

**Comment by wipfli** [comment]
It is nice to have the file hosted somewhere to test with mobile devices. And it is also convenient...

**Comment by HarelM** [comment]
There are basically two different features here I think.
The ability to add elevation to marker images, tec (original post) and the ability to add elevation to the geometry.
I'll see if I can split those two and maybe assign bounties to them as part of the 3D terrain effort.

**Comment by HarelM** [comment]
See the following design proposal in the spec repo:
https://github.com/maplibre/maplibre-gl-style-spec/issues/62
Feel free to join the discussion and suggest spec definitions to support this.

**Comment by HarelM** [comment]
While I think the demo repo here is great in terms of showing a proof of concept for 3D lines I think there's a problem with the underlaying source data in MVT format - MVT currently doesn't support elevation data so this need to be enriched in the client side somehow which is less than ideal.
Bottom line, I think it's possible but I think it will require a big change in the code in order to actually support this in a decent way without "hacks".
If someone would like to take a closer look at this feel free to do so.

**Comment by cns-solutions-admin** [comment]
See my comment https://github.com/maplibre/maplibre-style-spec/issues/62#issuecomment-1683603898

Though MVT currently does not support 3D geometries, maplibre could do so (at least) internally, thus supporting both 3D symbol and 3D line geojson layers.

**Comment by HarelM** [comment]
Another interesting approach is the use of "buildings" to draw lines that are above the ground.
See the demo in the below discussion.
https://github.com/maplibre/maplibre-gl-js/discussions/3397
cc: @wiesehahn

**Comment by cozzbp** [comment]
Is it possible for me to contribute to a public bounty for this feature?  Willing to pay someone to work on this.

**Comment by HarelM** [comment]
Yes, the problem is finding someone who is willing to work on it.
CC: @SnailBones.

You can post in the slack channel to see if there is an interest to getting paid for it.

**Comment by birkskyum** [comment]
@nyurik , is the mlt format flexible enough to support the 3d geometries that [mvt doesn't](https://github.com/maplibre/maplibre-gl-js/issues/644#issuecomment-1475377117)?

**Comment by jo-chemla** [comment]
The issue referenced by OP just [references](https://github.com/mapbox/mapbox-gl-js/issues/3993#issuecomment-2533717645) early release of a `fill-z-offset` property, with examples upcoming - altitude being defined relative to ground or sea level references.

**Comment by NeluQi** [comment]
This is very necessary!

**Comment by NeluQi** [comment]
When will this be available in the main package?

**Comment by HarelM** [comment]
When someone will invest the time to implement it.

**Comment by HarelM** [comment]
I've created the following branch:
https://github.com/maplibre/maplibre-gl-js/tree/symbol-elevation
Which adds support for symbol-elevation, but as an offset from terrain and not as a "override".
It doesn't do line, as those are a bit more complicated with their 3D "shape" is not really defined.
3D lines can be achieved to some extend using building extrusions.

https://github.com/user-attachments/assets/85eff05a-d20d-49b1-8167-e5afb63795ee

**Comment by clement-igonet** [comment]
Small prototype of per-level markers, DOM labels pinned to a feature altitude, including underground levels:

![per-level markers, floors -2 to 3](https://raw.githubusercontent.com/clement-igonet/maplibre-gl-js/assets/644-per-level-markers.png)

https://github.com/user-attachments/assets/656db793-7c7a-4057-8e92-c6e3a60223c8

Live: https://maplibre.confinia.io/debug/per-level-markers-644.html

Each label projects lngLat + altitude through the map's projection matrix, about 20 lines in userland, but it needs internals (map._camera.transform.modelViewProjectionMatrix), and the underground floors render thanks to #8069. A public project-with-altitude API (#8053) would make the pattern possible without touching internals.
